src/chains/ethereum/ethereum/src/api.ts

```diff
@@ -2884,6 +2884,198 @@ export default class EthereumApi implements Api {
       overrides
     );
   }
+
+  /**
+   * Returns a collection of historical block gas data and, optional, effective fee spent per unit of gas for a given percentile of block gas usage.
+   * Added in [EIP-1559](https://www.alchemy.com//blog/eip-1559)
+   *
+   * @param blockCount - Range of blocks between 1 and 1024. Will return less than the requested range if not all blocks are available.
+   * @param newestBlock - Highest block of the requested range.
+   * @param rewardPercentiles - (Optional) A monotonically increasing list of percentile values. For each block in the requested range,
+   * the transactions will be sorted in ascending order by effective tip per gas and the corresponding effective tip for the percentile
+   * will be determined, accounting for gas consumed.
+   * @returns transaction base fee per gas and effective priority fee per gas for the requested/supported block range
+   *
+   * * `oldestBlock`:  - Lowest number block of the returned range.
+   * * `baseFeePerGas`:  - An array of block base fees per gas. This includes the next block after the newest of the returned range,
+   * because this value can be derived from the newest block. Zeroes are returned for pre-EIP-1559 blocks.
+   * * `gasUsedRatio`:  - An array of block gas used ratios. These are calculated as the ratio of `gasUsed` and `gasLimit`.
+   * * `reward`:  - An array of effective priority fee per gas data points from a single block. All zeroes are returned if the
+   * block is empty.
+   *
+   * @example
+   * ```javascript
+   * const [from, to] = await provider.request({ method: "eth_accounts", params: [] });
+   * await provider.request({ method: "eth_sendTransaction", params: [{ from, to }] });
+   * const feeHistory = await provider.request({ method: "eth_feeHistory", params: ["0x1", "0x1", [10, 100]] });
+   * console.log(feeHistory);
+   * ```
+   */
+  @assertArgLength(3)
+  async eth_feeHistory(
+    blockCount: QUANTITY,
+    newestBlock: QUANTITY | Ethereum.Tag,
+    rewardPercentiles: number[]
+  ): Promise<Ethereum.FeeHistory<"private">> {
+    const blockchain = this.#blockchain;
+    const MIN_BLOCKS: number = 1;
+    const MAX_BLOCKS: number = 1024;
+    const PRECISION_FLOAT: number = 1e14;
+    const PAD_PRECISION = 16;
+    const PRECISION_BIG_INT: bigint = BigInt(1e16);
+
+    const newestBlockNumber = blockchain.blocks
+      .getEffectiveNumber(newestBlock)
+      .toBigInt();
+
+    // blockCount > newestBlock is technically valid but we cannot go past the Genesis Block.
+    // blockCount must be between 1 and 1024
+    const totalBlocks = Math.min(
+      Math.min(
+        Math.max(Quantity.toNumber(blockCount), MIN_BLOCKS),
+        Quantity.toNumber(newestBlockNumber + 1n)
+      ),
+      MAX_BLOCKS
+    );
```

Comment:
Math.min is variadic, so `Math.min(arg1, arg2, arg3)` could be used here.

Only potential problems I see here is that:

a) we clamp from 1-1024 instead of returning an error; if this is intentional, which IIRC it is, add a clarifying comment stating why.

b) `newestBlockNumber` _can_ be greater than `MAX_SAFE_INTEGER`. So while it is highly unlikely, we should probably handle it. :cry:

Reply:
:+1:
https://github.com/trufflesuite/ganache/pull/3351/commits/e3cf204a7281543b6f36f0c7ab6f4bd11570ea7a
https://github.com/trufflesuite/ganache/pull/3351/commits/91ebb6025b00a4ff559515c45f75ce6ba87487dc